Variable size transfer of quadrature point data implemented

rezarastak opened the pull request:
Fixes #8506 

This PR is not complete right now, it seems to work though. Looking forward to your feedback.

Progress:
- [x] `ContinuousQuadratureDataTransfer` transfers variables size quadrature data
- [x] Passes all existing tests.
- [ ] Properly handle the case where some cells don't need to store any data ([this comment](https://github.com/dealii/dealii/issues/8506#issuecomment-519935629))
- [ ] Add a new test.

 @jppelteret @davydden @marcfehling @peterrum

davydden commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -732,18 +732,17 @@ pack_cell_data(const CellIteratorType &                           cell,
 
-  const unsigned int n = matrix_data.n();
+  const auto m = qpd.size();
+  const auto n = qpd[0]->number_of_values();
```
should assert that m>0

marcfehling replied:
...or change both variables `m` and `n` to be of type `const unsigned int`.

I would suggest to refrain from using `auto` here since `unsigned int` makes the code more easily readable.

jppelteret replied:
I believe that what @davydden is suggesting is that `qpd[0]` actually exists. So `Assert(m>0)` does seem appropriate here.

marcfehling commented:
Nice job! There were really just a few lines that had to changed :-)
I'm curious about the test you are about to submit!

rezarastak commented:
This PR is ready to be reviewed. It has all the required new functionality and has appropriate new tests.

peterrum commented on tests/base/quadrature_point_data_04.cc:
```diff
@@ -0,0 +1,247 @@
+      else if (center[0] < 0.5 && center[1] >= 0.5)
+        cell->recursively_set_material_id(2);
+      else
+        // material_id 3 does not have an associated data structure
```
You can remove the comment here.

peterrum commented on tests/base/quadrature_point_data_04.cc:
```diff
@@ -0,0 +1,247 @@
+          storage.template initialize<Mat1>(cell, n_data_points_per_cell);
+        else if (cell->material_id() == 2)
+          storage.template initialize<Mat2>(cell, n_data_points_per_cell);
+      }
```
Move the comment `// material_id 3 does not have an associated data structure` here.

peterrum commented on tests/base/quadrature_point_data_04.cc:
```diff
@@ -0,0 +1,247 @@
+  initialize_data(tria, storage);
+
+  // Assign some value to the cell data
+  FE_Q<2>     fe(1);
```
Why is `degree` not  a constexpr, but `n_quadrature_per_dim`?

rezarastak replied:
I removed the consexpr global variable and made it more consistent.

peterrum commented on tests/base/quadrature_point_data_04.cc:
```diff
@@ -0,0 +1,247 @@
+      {
+        // It does not modify the data on already initialized cell data
+        if (cell->material_id() == 0)
+          storage.template initialize<Mat0>(cell, n_data_points_per_cell);
```
I would instead of using `n_data_points_per_cell` use `quad->size()`

peterrum commented on tests/base/quadrature_point_data_04.cc:
```diff
@@ -0,0 +1,247 @@
+  }
+};
+
+constexpr unsigned int n_quadrature_per_dim = 3;
```
I would move this line after the includes.

peterrum commented on tests/base/quadrature_point_data_04.cc:
```diff
@@ -0,0 +1,247 @@
+};
+
+constexpr unsigned int n_quadrature_per_dim = 3;
+constexpr unsigned int n_data_points_per_cell =
```
I would remove this line; see also other comments.

bangerth commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -707,6 +713,14 @@ CellDataStorage<CellIteratorType, DataType>::get_data(
+template <typename CellIteratorType, typename DataType>
+bool
+CellDataStorage<CellIteratorType, DataType>::stores_data(
+  const CellIteratorType &cell) const
```
You might want to mark the function as `inline` since it lives in a header file.

rezarastak replied:
All of the function definitions in this file don't explicitly contain the `inline` keyword but they are all implicitly inline because they are templated.

bangerth replied:
That's not correct. `template` functions have weak linkage, but they're not automatically `inline`.

rezarastak replied:
Okay! Do you think other function definitions in this header file need `inline` keyword as well? or is it just this `stores_data()` method because it is small?

bangerth replied:
In principle, every function that is small enough that inlining makes sense. That requires inspection of every function defined in the .h file.

bangerth commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -168,6 +168,12 @@ class CellDataStorage : public Subscriptor
+   * Queries whether a data belonging to @p cell exists.
+   */
+  bool
+  stores_data(const CellIteratorType &cell) const;
```
It's a shame that we didn't anticipate the fact that some cells might not have data associated with them, or we could have let `get_data()` return a `boost::optional` object.

rezarastak replied:
I also think changing the output of `get_data()` to give additional info on whether a data exists would have been a much better design. We discussed it in #8506 and decided to go with this design (adding member function `stores_data()` instead).

I prefer writing
```c++
auto qpd = data_storage(cell);
if (!qpd)
  return;
```
instead of the current expression
```c++
if (!data_storage->stores_data(cell))
    return;
auto qpd = data_storage->get_data(cell);
```
because the latter requires two map lookups instead of just one lookup.
However, backward compatibility limits us a bit here because it forces us to choose the second design.

bangerth replied:
Or do `if (auto data = data_storage->get_data) { ... }` :-)

Like I mentioned elsewhere, we always have the option of deprecating `get_data()` and invent a new function that has the right interface.

jppelteret replied:
I'm happy if you want to follow through on @bangerth's suggestion if it seems like the better solution to the majority of people. @bangerth Maybe you wouldn't mind reviewing this again once its been updated -- I'm still short on time ATM.

bangerth replied:
@jppelteret -- yes, gladly!
@rezarastak -- let me know once you want me to take another look!

bangerth commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -729,21 +743,27 @@ pack_cell_data(const CellIteratorType &                           cell,
     std::is_base_of<TransferableQuadraturePointData, DataType>::value,
     "User's DataType class should be derived from QPData");
 
+  if (!data_storage->stores_data(cell))
```
In most places, we use this instead:
```suggestion
  if (data_storage->stores_data(cell) == false)
```

bangerth commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -729,21 +743,27 @@ pack_cell_data(const CellIteratorType &                           cell,
 
+  if (!data_storage->stores_data(cell))
+    {
+      matrix_data = FullMatrix<double>();
```
```suggestion
      matrix_data.clear();
```

rezarastak replied:
I couldn't find the `clear()` method in the `FullMatrix<Number>` interface. I used the `reinit({0, 0})` method instead.

bangerth replied:
Yes, that makes sense.

bangerth commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -729,21 +743,27 @@ pack_cell_data(const CellIteratorType &                           cell,
+      AssertDimension(single_qp_data.size(), n);
 
-      for (unsigned int i = 0; i < n; i++)
+      for (unsigned int i = 0; i < n; ++i)
```
:-)

bangerth commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -763,17 +783,20 @@ unpack_to_cell_data(const CellIteratorType &                     cell,
     std::is_base_of<TransferableQuadraturePointData, DataType>::value,
     "User's DataType class should be derived from QPData");
 
+  if (!data_storage->stores_data(cell))
```
same here

marcfehling commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -897,10 +895,11 @@ namespace parallel
 
-      // to get consistent data sizes on each cell for the fixed size transfer,
-      // we won't allow compression
       return Utilities::pack(matrix_dofs, /*allow_compression=*/false);
```
Please consider introducing an optimized pack function.

marcfehling commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -925,11 +924,18 @@ namespace parallel
         Utilities::unpack<FullMatrix<double>>(data_range.begin(),
```
Please consider introducing an optimized unpack function.

rezarastak commented:
@marcfehling I suggest performing this pack/unpack optimization at a separate PR in the future. We first have to use this variable quadrature data transfer feature and benchmark its performance before deciding whether we want to specialize the pack/unpack functions.

bangerth commented:
/rebuild

bangerth commented:
I agree with the approach of doing one thing at a time.

rezarastak commented:
Sorry @bangerth, I haven't yet implemented the modified `get_data()` method that returns an optional yet. Give me another day or two to finish that before merging this.

masterleinad commented:
Just tell us when you are ready and we should have another look.

rezarastak commented:
I added the function `try_get_data()` that returns an optional if data is not available. I decided to not deprecate `get_data()` because it is useful as it provides a simpler interface for the cases where we are sure that data exists for all cells. For example, step-44 would have become unnecessarily complicated if I replaced `get_data()` with `try_get_data()`.

This PR is ready to be reviewed/tested.

masterleinad commented on doc/news/changes/minor/20190821RezaRastak:
```diff
@@ -0,0 +1,5 @@
+New: method try_get_data() added to CellDataStorage in both const and
+non-const forms. It returns an optional which determines whether a
```
```suggestion
non-const forms. It returns an std_cxx17::optional which determines whether a
```

masterleinad commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -168,6 +169,44 @@ class CellDataStorage : public Subscriptor
   get_data(const CellIteratorType &cell) const;
 
+  /**
+   * Returns an optional indicating whether @p cell contains an associated
```
```suggestion
   * Returns an std_cxx17::optional indicating whether @p cell contains an associated
```

masterleinad commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -168,6 +169,44 @@ class CellDataStorage : public Subscriptor
 
+  /**
+   * Returns an optional indicating whether @p cell contains an associated
+   * data or not. If data is available, dereferencing the optional
```
```suggestion
   * data or not. If data is available, dereferencing the std_cxx17::optional
```

masterleinad commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -168,6 +169,44 @@ class CellDataStorage : public Subscriptor
+  try_get_data(const CellIteratorType &cell);
+
+  /**
+   * Returns an optional indicating whether @p cell contains an associated
```
```suggestion
   * Returns an std_cxx17::optional indicating whether @p cell contains an associated
```

masterleinad commented on include/deal.II/base/quadrature_point_data.h:
```diff
@@ -168,6 +169,44 @@ class CellDataStorage : public Subscriptor
+
+  /**
+   * Returns an optional indicating whether @p cell contains an associated
+   * data or not. If data is available, dereferencing the optional reveals
```
```suggestion
   * data or not. If data is available, dereferencing the std_cxx17::optional reveals
```
